File: frontend/middleware.ts
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";
import { v4 as uuidv4 } from "uuid";

export function middleware(request: NextRequest) {
    let response = NextResponse.next();

    // Retrieve cart_id cookie
    const cartId = request.cookies.get("cart_id");

    if (!cartId) {
        const newCartId = uuidv4();
        response.cookies.set("cart_id", newCartId, {
            httpOnly: true,
            secure: process.env.NODE_ENV === "production",
            maxAge: 365 * 24 * 60 * 60, // 1 year
            path: "/",
            sameSite: "lax",
        });
    }

    return response;
}

export const config = {
    matcher: ["/", "/login", "/register", "/:sku/:slug"],
};

File: backend/order/middleware.py
import uuid
from django.utils.deprecation import MiddlewareMixin
from django.conf import settings





class CartIDMiddleware(MiddlewareMixin):
    """
    Middleware to ensure every user (guest or authenticated) has a `cart_id` cookie.
    This prevents errors when handling carts.
    """

    def process_response(self, request, response):
        if "cart_id" not in request.COOKIES:
            cart_id = str(uuid.uuid4())
            response.set_cookie(
                key="cart_id",
                value=cart_id,
                max_age=365 * 24 * 60 * 60,  # 1 year
                path="/",
                secure=not settings.DEBUG,  # Secure in production
                httponly=True,
                samesite="Lax",
            )
        return response